File: plugins/wise/engine/src/phases/worktree.ts
// `worktree`: bring `<runDir>/worktrees/<slug>` into being on the unit's branch, from a freshly
// fetched base (process-tickets.md §1 "Ensure the worktree"): reuse a registered worktree on our
// branch, re-attach a bare branch, prune a stale entry, else create branch and worktree together.
// `.worktreeinclude` files are copied once per worktree, gated by the ledger.

import { existsSync, mkdirSync, readdirSync, rmdirSync } from "node:fs";
import { dirname } from "node:path";
import { applyWorktreeInclude } from "../ledger.ts";
import { realpathLoose } from "../paths.ts";
import {
  errText,
  fail,
  git,
  localBranchExists,
  NETWORK_CMD_TIMEOUT_MS,
  ok,
  pass,
} from "./common.ts";
import type { PhaseCtx, PhaseResult, PhaseRunner } from "./common.ts";

/** Ledger marker: `cursors.worktree` once the include files were copied. */
export const INCLUDES_DONE = "includes-done";

type Registered = { path: string; branch?: string };

/** Parse `git worktree list --porcelain` into path + branch pairs. */
export function parseWorktrees(porcelain: string): Registered[] {
  const out: Registered[] = [];
  for (const block of porcelain.split(/\n\n+/)) {
    let path: string | undefined;
    let branch: string | undefined;
    for (const line of block.split("\n")) {
      if (line.startsWith("worktree ")) path = line.slice("worktree ".length);
      else if (line.startsWith("branch refs/heads/"))
        branch = line.slice("branch refs/heads/".length);
    }
    if (path !== undefined) out.push(branch !== undefined ? { path, branch } : { path });
  }
  return out;
}

async function registered(ctx: PhaseCtx, path: string): Promise<Registered | undefined> {
  const r = await git(ctx, ["worktree", "list", "--porcelain"]);
  if (!ok(r)) return undefined;
  const want = realpathLoose(path);
  return parseWorktrees(r.stdout).find((w) => realpathLoose(w.path) === want);
}

function dirIsEmpty(path: string): boolean {
  try {
    return readdirSync(path).length === 0;
  } catch {
    return false;
  }
}

export const worktreePhase: PhaseRunner = async (ctx): Promise<PhaseResult> => {
  const { unit } = ctx;
  const path = unit.worktree;
  const base = unit.base || "main";

  const fetched = await git(ctx, ["fetch", "origin", base], { timeoutMs: NETWORK_CMD_TIMEOUT_MS });
  if (!ok(fetched)) {
    const have = await git(ctx, ["rev-parse", "--verify", "--quiet", `origin/${base}`]);
    if (!ok(have)) return fail(`worktree: fetch origin/${base} failed and no local copy`);
    ctx.log(`worktree: fetch failed, using the local origin/${base}`);
  }

  let reg = await registered(ctx, path);
  if (reg === undefined && existsSync(path)) {
    // Directory without an admin entry: stale or corrupt. Prune, then re-attach an empty dir.
    await git(ctx, ["worktree", "prune"]);
    reg = await registered(ctx, path);
    if (reg === undefined) {
      if (!dirIsEmpty(path)) return fail(`worktree-corrupt: ${path} exists but is not a worktree`);
      rmdirSync(path);
    }
  }
  if (reg !== undefined) {
    if (reg.branch !== unit.branch) {
      return fail(`worktree-conflict: ${path} is on ${reg.branch ?? "detached HEAD"}`);
    }
    ctx.log(`worktree: reuse ${path}`);
  } else {
    mkdirSync(dirname(path), { recursive: true });
    const add = (await localBranchExists(ctx, unit.branch))
      ? await git(ctx, ["worktree", "add", path, unit.branch])
      : await git(ctx, [
          "worktree",
          "add",
          "--no-track",
          path,
          "-b",
          unit.branch,
          `origin/${base}`,
        ]);
    if (!ok(add)) return fail(`worktree: git worktree add failed: ${errText(add)}`);
    ctx.log(`worktree: created ${path} on ${unit.branch}`);
  }

  const cursors = { ...ctx.ledger.cursors };
  if (cursors.worktree !== INCLUDES_DONE) {
    const inc = applyWorktreeInclude(ctx.cwd, path);
    for (const n of inc.notices) ctx.log(n);
    if (inc.copied > 0) ctx.log(`worktree: copied ${inc.copied} include path(s)`);
    cursors.worktree = INCLUDES_DONE;
  }
  return pass({ unit: { ...unit, worktree: path, base }, cursors });
};

File: plugins/wise/engine/wise_engine/phases/worktree.py
import re
from pathlib import Path

from ..ledger import apply_worktree_include
from .common import (
    Json,
    NETWORK_CMD_TIMEOUT_MS,
    base_ref,
    err_text,
    fail,
    git,
    local_branch_exists,
    ok,
    pass_,
)
from .remote import remote_of

INCLUDES_DONE = "includes-done"


def parse_worktrees(porcelain: str) -> list[Json]:
    rows = []
    for block in re.split(r"\n\n+", porcelain):
        row = {}
        for line in block.split("\n"):
            if line.startswith("worktree "):
                row["path"] = line[len("worktree ") :]
            elif line.startswith("branch refs/heads/"):
                row["branch"] = line[len("branch refs/heads/") :]
        if "path" in row:
            rows.append(row)
    return rows


async def _registered(ctx: Json, path: str) -> Json | None:
    result = await git(ctx, ["worktree", "list", "--porcelain"])
    if not ok(result):
        return None
    want = Path(path).resolve()
    return next(
        (row for row in parse_worktrees(result["stdout"]) if Path(row["path"]).resolve() == want),
        None,
    )


async def worktree_phase(ctx: Json) -> Json:
    unit = ctx["unit"]
    path = Path(unit["worktree"])
    base = unit["base"] or "main"
    remote = remote_of(ctx)
    fetched = (
        None
        if remote["kind"] == "none"
        else await git(ctx, ["fetch", "origin", base], {"timeout_ms": NETWORK_CMD_TIMEOUT_MS})
    )
    ref = await base_ref(ctx, base)
    if ref is None:
        return fail(f"worktree: base {base} exists neither on origin nor locally")
    # A GitHub PR cannot target a branch origin does not have; without a
    # GitHub remote no PR is opened, so a local-only base is fine.
    if remote["kind"] == "github" and not ref.startswith("origin/"):
        return fail(f"worktree: base {base} exists only locally; push it to origin first")
    if fetched is not None and not ok(fetched):
        ctx["log"](f"worktree: fetch origin {base} failed, using the last fetched {ref}")
    unit = {**unit, "base_ref": ref}
    if path.resolve() == Path(ctx["cwd"]).resolve():
        head = await git(ctx, ["symbolic-ref", "--quiet", "--short", "HEAD"])
        if not ok(head) or head["stdout"].strip() != unit["branch"]:
            status = await git(ctx, ["status", "--porcelain"])
            if not ok(status) or status["stdout"].strip():
                return fail("worktree: current tree has uncommitted or untracked changes")
            args = (
                ["checkout", unit["branch"]]
                if await local_branch_exists(ctx, unit["branch"])
                else ["checkout", "--no-track", "-b", unit["branch"], ref]
            )
            switched = await git(ctx, args)
            if not ok(switched):
                return fail(f"worktree: checkout failed: {err_text(switched)}")
        return pass_({"unit": {**unit, "worktree": str(path), "base": base}})
    reg = await _registered(ctx, str(path))
    if reg is None and path.exists():
        await git(ctx, ["worktree", "prune"])
        reg = await _registered(ctx, str(path))
        if reg is None:
            try:
                empty = not list(path.iterdir())
            except OSError:
                empty = False
            if not empty:
                return fail(f"worktree-corrupt: {path} exists but is not a worktree")
            path.rmdir()
    if reg is not None:
        if reg.get("branch") != unit["branch"]:
            return fail(f"worktree-conflict: {path} is on {reg.get('branch', 'detached HEAD')}")
        ctx["log"](f"worktree: reuse {path}")
    else:
        path.parent.mkdir(parents=True, exist_ok=True)
        args = (
            ["worktree", "add", str(path), unit["branch"]]
            if await local_branch_exists(ctx, unit["branch"])
            else [
                "worktree",
                "add",
                "--no-track",
                str(path),
                "-b",
                unit["branch"],
                ref,
            ]
        )
        added = await git(ctx, args)
        if not ok(added):
            return fail(f"worktree: git worktree add failed: {err_text(added)}")
        ctx["log"](f"worktree: created {path} on {unit['branch']}")
    cursors = dict(ctx["ledger"]["cursors"])
    if cursors.get("worktree") != INCLUDES_DONE:
        result = apply_worktree_include(ctx["cwd"], str(path))
        for notice in result["notices"]:
            ctx["log"](notice)
        if result["copied"] > 0:
            ctx["log"](f"worktree: copied {result['copied']} include path(s)")
        cursors["worktree"] = INCLUDES_DONE
    return pass_({"unit": {**unit, "worktree": str(path), "base": base}, "cursors": cursors})
